os/pm/pm_idle.c

```diff
@@ -0,0 +1,104 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+#include <assert.h>
+#include <tinyara/pm/pm.h>
+#include <tinyara/clock.h>
+#include <tinyara/irq.h>
+#include <tinyara/arch.h>
+
+#include "pm.h"
+
+#ifdef CONFIG_PM
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: pm_idle
+ *
+ * Description:
+ *   This function is called by IDLE thread to make board sleep. This function
+ *   also allow to set wake up timer & handler and do all the PM pre processing
+ *   required before going to sleep.
+ *
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+void pm_idle(void)
+{
+	irqstate_t flags;
+	enum pm_state_e newstate;
+#ifdef CONFIG_PM_TIMEDWAKEUP
+	clock_t delay;
+#endif
+	/* Decide, which power saving level can be obtained */
+	flags = enter_critical_section();
+	newstate = pm_checkstate();
+	/* Perform state-dependent logic here */
+	pmvdbg("newstate= %d\n", newstate);
+	/* Then force the global state change */
+	if (pm_changestate(newstate) < 0) {
+		/* The new state change failed */
+		pmdbg("State change failed! newstate = %d\n", newstate);
+		/* core power efficiency during idle time */
+		leave_critical_section(flags);
+		up_pm_core_sleep();
+		return;
+	}
+	/* If current state is not good to go sleep then do core power saving*/
+	if (g_pmglobals.state != PM_SLEEP) {
+		/* core power efficiency during idle time */
+		leave_critical_section(flags);
+		up_pm_core_sleep();
+		return;
+	}
+#ifdef CONFIG_PM_TIMEDWAKEUP
+	/* Disable interrupts through out this operation to set wakeup timer */
+	delay = wd_getwakeupdelay();
+	if (delay >= CONFIG_PM_MIN_SLEEP_TIME) {
+		pmvdbg("Setting timer and Board will wake up after %d millisecond\n", delay);
+		up_set_pm_timer(TICK2USEC(delay));
+	} else {
+		pmvdbg("Min Sleep Time should be %d\n", CONFIG_PM_MIN_SLEEP_TIME);
```

Comment:
i think you should return here, because we dont want board to sleep if delay is very less

Reply:
As discussed with @ewoodev, for this case we are making the board sleep till Interrupts.

Reply:
sorry for confusing @gSahitya-samsung.
We need to check whether there is a delay or not and if there is no wakeup source.
If there is a wakeup source but the delay is smaller than CONFIG_PM_MIN_SLEEP_TIME, we should not enter sleep mode.
I made a wrong infomation.

if delay is 0, there is no wakeup source in wdog.
Please add check delay 0.
if delay == 0, sleep forever.